store/src/main/java/com/nytimes/android/external/store3/pipeline/PipelinePersister.kt

```diff
@@ -1,157 +1,138 @@
+                        is Command.Read -> read(command, request, sink)
+                        is Command.Stop -> command.ack.complete(Unit)
+                        Command.Fetch -> fetch(request, sink)
                     }
+
+                } catch (t: Throwable) {
+                    sink.send(StoreResponse.Error(t, ResponseOrigin.Persister))
                 }
             }
         }
-        diskCommands.consumeAsFlow().collectLatest { command ->
-            when (command) {
-                is DiskCommand.Stop -> {
-                    command.ack.complete(Unit)
-                }
-                is DiskCommand.ReadFirst -> {
-                    reader(request.key).collectIndexed { index, diskData ->
-                        diskData?.let {
-                            send(
-                                StoreResponse.Data(
-                                    value = diskData,
-                                    origin = ResponseOrigin.Persister
-                                )
-                            )
-                        }
 
-                        if (index == 0 && (diskData == null || request.refresh)) {
-                            fetcherCommands.send(
-                                FetcherCommand.Fetch
-                            )
-                        }
-                    }
-                }
-                is DiskCommand.Read -> {
-                    var fetcherOrigin : ResponseOrigin? = command.origin
-                    reader(request.key).collect { diskData ->
-                        if (diskData != null) {
-                            val origin = fetcherOrigin?.let {
-                                fetcherOrigin = null
-                                it
-                            } ?: ResponseOrigin.Persister
-                            send(
-                                StoreResponse.Data(
-                                    value = diskData,
-                                    origin = origin
-                                )
-                            )
-                        }
-                    }
+        if (request.shouldSkipCache(CacheType.DISK)) {
+            actor.offer(Command.Fetch)
+        } else {
+            actor.offer(Command.ReadFirst)
+        }
+        return actor
+    }
+
+    private suspend fun ActorScope<Command>.fetch(request: StoreRequest<Key>, sink: SendChannel<StoreResponse<Output>>) {
+        fetcher.stream(request).collect {
+            val data = it.dataOrNull()
+            if (data != null) {
+                try {
+                    // stop disk first
+                    val ack = CompletableDeferred<Unit>()
+                    channel.send(Command.Stop(ack))
+                    ack.await()
+                    writer(request.key, data)
+                } finally {
+                    channel.send(Command.Read(it.origin))
                 }
+            } else {
+                sink.offer(it.swapType<Output>())
```

Comment:
the sink that is being passed down here is unlimited.
taht means this will keep asking for more values from the fetcher  as soon as sink.offer returns (which it will immediataely).
I think that is a problem, this should behave like a flow (or Rx2 Flowable) where we don't ask for more values unless we are asked more values.

Reply:
open to suggestions I was not familiar with the capacities and what effect they have